### PR TITLE
Support LeakyReluGrad for 5D tensors in Layout optimizer

### PR DESCRIPTION
**Pull request by kaixih**
This PR is to allow the grappler layout optimizer to support unary grad ops (e.g., LeakyReluGrad) for 5D tensors.

fyi. @nluehr

### CODE REVIEW

File: tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc [hunk 1]
```diff
@@ -1797,13 +1797,39 @@ Status TileTransposer::TransposeNode(TransposeContext* context,
 Status UnaryGradTransposer::TransposeNode(TransposeContext* context,
                                           utils::MutableNodeView* node) {
   DCHECK(IsUnaryGrad(*node->node()));
-  if (!ShouldProcess(*context, *node) || !IsFanoutPortRankN(*node, 0, 4) ||
+  const auto* output_shape_attr = node->GetAttr(kAttrOutputShape);
+  const auto& shape = output_shape_attr->list().shape(0);
+  const int rank = shape.dim_size();
+  std::string src_format = context->src_format;
```
Comment (andyly) [inline tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc:1803]: Can `ScopedDataFormatUpgrader` (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc#L173) be used instead? https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc#L719-L736 is an example of it being used.

**Review by andyly — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
@kaixih  Can you please check @andyly's comments and keep us posted ? Thanks!

**Review by andyly — APPROVED** [review]

File: tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc [hunk 2]
```diff
@@ -1797,10 +1797,17 @@ Status TileTransposer::TransposeNode(TransposeContext* context,
 Status UnaryGradTransposer::TransposeNode(TransposeContext* context,
                                           utils::MutableNodeView* node) {
   DCHECK(IsUnaryGrad(*node->node()));
-  if (!ShouldProcess(*context, *node) || !IsFanoutPortRankN(*node, 0, 4) ||
+  const auto* output_shape_attr = node->GetAttr(kAttrOutputShape);
+  const auto& shape = output_shape_attr->list().shape(0);
+  const int rank = shape.dim_size();
```
Comment (andyly) [inline tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc:1802]: Can this be replaced with using `GetFanoutPortRank` (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc#L736-L739) along with an explicit rank size check? Currently this does not restrict the transposer to only allow for rank 4 and rank 5 tensors.

File: tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc [hunk 3]
```diff
@@ -1797,10 +1797,17 @@ Status TileTransposer::TransposeNode(TransposeContext* context,
...
+  ScopedDataFormatUpgrader data_format_upgrader(context, rank);
+  if (!ShouldProcess(*context, *node) || !IsFanoutPortRankN(*node, 0, rank) ||
```
Comment (andyly) [inline tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc:1804]: The `IsFanoutPortRankN` check is not necessary after the change above.

**Review by andyly — CHANGES_REQUESTED** [review]

File: tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc (hunk 3 above)
Comment (kaixih) [reply tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc]: Yes, I see, since rank is already obtained from the 0th fanout port.

**Comment by kaixih** [comment]
@andyly PTAL.

**Comment by andyly** [comment]
> @andyly PTAL.

Can you also address the other comment for explicitly checking the rank? The transposer originally checked for if the output is rank 4 but now there is no check or restriction of rank. You should be able to replace getting and checking the rank with something like https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/optimizers/generic_layout_optimizer_transposer.cc#L736-L739.

**Comment by kaixih** [comment]
Sure. I guess I missed that comments. Changes are made. PTAL.

**Review by andyly — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Pull request merged**